From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nicolas Bouchinet <nicolas.bouchinet@ssi.gouv.fr>
Date: Mon, 23 Mar 2026 09:47:13 +0100
Subject: [PATCH 102/103] slub: Add canary_debug kernel parameter

With canary_debug, a canary mismatch will print supposed canary values
and the one that has been encountered.

Signed-off-by: Levente Polyak <levente@leventepolyak.net>
Signed-off-by: Nicolas Bouchinet <nicolas.bouchinet@ssi.gouv.fr>
---
 mm/slub.c | 37 ++++++++++++++++++++++++++++++++++++-
 1 file changed, 36 insertions(+), 1 deletion(-)

diff --git a/mm/slub.c b/mm/slub.c
index 8caa30508fec0..fcb4e8c1c5e68 100644
--- a/mm/slub.c
+++ b/mm/slub.c
@@ -925,14 +925,49 @@ static inline void set_canary(struct kmem_cache *s, void *object, unsigned long
 {
 	unsigned long *canary = get_canary(s, object);
 	*canary = get_canary_value(canary, value);
 }
 
+static inline void print_canary_value(struct kmem_cache *s, void * object, unsigned long value)
+{
+	unsigned long *canary = get_canary(s, object);
+
+	early_printk("check_canary: canary mismatch on cache (%s) "
+	      "for object %p:\n"
+	      "\tchecked canary value = %lx\n"
+	      "\tobject canary value = %lx\n"
+	      "\tpossible canary values for the cache :\n"
+	      "\trandom_active = %lx\n"
+	      "\trandom_inactive = %lx\n"
+	      "\tsheaf_random_inactive = %lx\n",
+	      s->name,
+	      object,
+	      get_canary_value(canary, value),
+	      *canary,
+	      get_canary_value(canary, s->random_active),
+	      get_canary_value(canary, s->random_inactive),
+	      get_canary_value(canary, s->sheaf_random_inactive));
+}
+
+static bool canary_debug __ro_after_init = false;
+static int __init setup_canary_debug(char *str)
+{
+	canary_debug = true;
+	return 1;
+}
+__setup_param("canary_debug", canary_debug, setup_canary_debug, 0);
+__setup("canary_debug", setup_canary_debug);
+
 static inline void check_canary(struct kmem_cache *s, void *object, unsigned long value)
 {
 	unsigned long *canary = get_canary(s, object);
-	BUG_ON(*canary != get_canary_value(canary, value));
+
+	if (*canary != get_canary_value(canary, value)) {
+		if (unlikely(canary_debug))
+			print_canary_value(s, object, value);
+		BUG_ON(1);
+	}
 }
 
 static inline void check_set_canary(struct kmem_cache *s, void *object, unsigned long check_value, unsigned long set_value)
 {
 	if (object && !is_kfence_address(object)) {
-- 
2.55.0

